backvendor/vendored.go

```diff
@@ -159,47 +160,66 @@ func updateHashesAfterStrip(hashes *FileHashes, wt *WorkingTree, ref string, pat
 	return anyChanged, nil
 }
 
-func matchFromRefs(strip bool, hashes *FileHashes, wt *WorkingTree, refs []string) (string, error) {
+func matchFromRefs(strip bool, hashes *FileHashes, wt *WorkingTree, refs []string) ([]string, error) {
 	var paths []string
 	if strip {
 		for hash, _ := range hashes.hashes {
 			paths = append(paths, hash)
 		}
 	}
 
-Refs:
+	matchFromRef := func(th *FileHashes, ref string) (bool, error) {
+		if hashes.IsSubsetOf(th) {
+			return true, nil
+		}
+
+		if !strip {
+			return false, nil
+		}
+
+		for _, path := range paths {
+			if _, ok := th.hashes[path]; !ok {
+				// File missing from revision
+				return false, nil
+			}
+		}
+
+		changed, err := updateHashesAfterStrip(th, wt, ref, paths)
+		if err != nil {
+			return false, err
+		}
+
+		return changed && hashes.IsSubsetOf(th), nil
+	}
+
+	matches := make([]string, 0)
 	for _, ref := range refs {
 		log.Debugf("%s: trying match", ref)
 		tagHashes, err := wt.FileHashesFromRef(ref)
```

Comment:
Optional: It might be better to name this `refHashes` because the refs accepted in `matchFromRefs` don't have to be tags.